dask/array/core.py

```diff
@@ -2833,20 +2833,21 @@ def where(condition, x=None, y=None):
     if x is None or y is None:
         raise TypeError(where_error_message)
 
-    x = asarray(x)
-    y = asarray(y)
-
-    shape = broadcast_shapes(x.shape, y.shape)
-    dtype = np.promote_types(x.dtype, y.dtype)
+    if np.isscalar(condition):
+        # Match the dtype inference logic from elemwise
+        vals = [np.empty((1,) * a.ndim, dtype=a.dtype)
+                if not is_scalar_for_elemwise(a) else a
+                for a in (x, y)]
+        dtype = apply_infer_dtype(np.where, [condition] + vals, {}, 'where', suggest_dtype=False)
```

Comment:
I'd had the same thought. What do you think about calling this function `result_type` rather than `infer_dtype` and decorating it as `@wraps(np.result_type)`?